### DISCUSSION TITLE
Inner products in pyMOR

### DISCUSSION THREAD

**Discussion by artpelling** [pymor/pymor#1743] [Q&A]
Hi all, I encountered some questions about inner products when taking a closer look at the randomized LA methods in pyMOR, maybe you could help me with that.

From what I understand, an `Operator` has a `source` and `range` `VectorSpace` attribute but the spaces are not associated with any inner products explicitly. In order to compute the inner product of two `VectorArrays`, `U` and `V`, belonging to, say, the `source` of an operator with respect to some `source_product`, we would have to pass that `source_product` to the `inner` method:

``` python
U.inner(V, product=source_product)
```
1. Do I understand correctly thus far?

2. From a mathematical standpoint, I would say that an inner product is a "property" of the vector space (e.g. if the vector space is a Hilbert space). What is the reasoning behind not having the product as an attribute to `VectorSpace` that defaults to the Euclidian inner product and remove the `product=None` arguments form methods such as `inner`? Technically, every vector space in pyMOR would then be a Hilbert space but isn't this the case now anyway since `inner` also defaults to the Euclidian inner product?

3. It seems inconsistent to me that `AdjointOperator` is the only operator with `source_product` and `range_product` attributes. If they were attributes of the `VectorSpaces` these attributes could be removed and the `apply` methods could be simplified. If that is not advisable, would you see merit in adding `source_product` and `range_product` attributes to all operators such that one wouldn't have to pass these around manually to all other functions like `gram_schmidt`?

4. The reason I am asking all this: In the randomized methods, we draw random samples of an operator's `source`. I was wondering whether the inner product should be respected by `source.random` per default such that
``` python
op.source.random(n, distribution='normal')
```
returns what would currently be returned by
``` python
source_product.apply(op.source.random(n, distribution='normal'))
```

ATM, I am not too sure about 4 but I think there might be some error/inconsistency in the `rand_la` module which I am trying to get to the bottom of. I don't see how the lines
``` python
    for i in range(q):
        Q = A.apply_adjoint(range_product.apply(Q))
        Q = source_product.apply_inverse(Q)
        gram_schmidt(Q, source_product, atol=0, rtol=0, copy=False)
        Q = A.apply(Q)
        gram_schmidt(Q, range_product, atol=0, rtol=0, copy=False)
```
in `rand_la.rrf` correspond to Algorithm 2 of the cited paper.

**Comment by sdrave** [comment]
Hi @artpelling,

>  1. Do I understand correctly thus far?

Yes!

>   2. From a mathematical standpoint, I would say that an inner product is a "property" of the vector space (e.g. if the vector space is a Hilbert space). 

Yes, many people have expressed their desire to have the inner product attached to the `VectorSpace` (me included).

> What is the reasoning behind not having the product as an attribute to `VectorSpace` that defaults to the Euclidian inner product and remove the `product=None` arguments form methods such as `inner`? Technically, every vector space in pyMOR would then be a Hilbert space but isn't this the case now anyway since `inner` also defaults to the Euclidian inner product?

There are several challenges: For instance, what do you do when you have multiple inner products around (say $H^1$ and $L^2$). If only one can be attached then the array would have to change space when you want to, e.g., orthogonalize w.r.t. the other product using `gram_schmidt`. (Or keep the cumbersome `product` argument.) If you have `VectorArrays` from the same (algebraic) space but with different inner products: what would their sum be? Mathematically most severe: if I write `project(A, W, V)` the meaning depends on whether `A` is actually a linear operator from a Hilbert space to itself, in which case we would need the take the inner products with the vectors in `W`, or a bilinear form, in which case `A` would map to the dual space and `project` would have to take the dual pairing. Currently, it is implicitly assumed in this case, that the result of `A.apply` is given w.r.t. to the dual basis s.t. the dual pairing would just be the Euclidean inner product of the DOF vectors. One motivation for attaching inner products to `VectorSpaces` would be that for most of pyMOR this would allow us to work completely 'coordinate-free', not making any assumptions about the basis representation of these vectors. (@lbalicki has a use case for this, for instance.) But then we would have to introduce the concept of duality. I actually started working on this, but ran out of time. Since then the question is in limbo.

>   3. It seems inconsistent to me that `AdjointOperator` is the only operator with `source_product` and `range_product` attributes.

I don't think this is inconsistent: You can only define the adjoint of an `Operator` when you know the inner products w.r.t. which you want to take the adjoint.

> If they were attributes of the `VectorSpaces` these attributes could be removed and the `apply` methods could be simplified. 

That's true.

> If that is not advisable, would you see merit in adding `source_product` and `range_product` attributes to all operators such that one wouldn't have to pass these around manually to all other functions like `gram_schmidt`?

I don't know about this. How would `source_product` and `range_product` affect the behavior of the operator's interface (except taking the adjoint)?


>   4. The reason I am asking all this: In the randomized methods, we draw random samples of an operator's `source`. I was wondering whether the inner product should be respected by `source.random` per default such that
> 
> 
> ```python
> op.source.random(n, distribution='normal')
> ```
> 
> returns what would currently be returned by
> 
> ```python
> source_product.apply(op.source.random(n, distribution='normal'))
> ```

I'm not sure if that makes sense. What properties would you like the result to have? I think what you would want is a standard normal distribution of the coefficients w.r.t. an orthonormal basis of the Hilbert space. For that you would need to compute the inverse square root of `source_product` if I'm not mistaken. Since this is not so easy, `adaptive_rrf` pays `1/sqrt(lambda_min)` as a penalty for using the wrong basis. As far as I remember, when @andreasbuhr worked on this, he could not find a way to resolve this issue.


> ATM, I am not too sure about 4 but I think there might be some error/inconsistency in the `rand_la` module which I am trying to get to the bottom of. I don't see how the lines
> 
> ```python
>     for i in range(q):
>         Q = A.apply_adjoint(range_product.apply(Q))
>         Q = source_product.apply_inverse(Q)
>         gram_schmidt(Q, source_product, atol=0, rtol=0, copy=False)
>         Q = A.apply(Q)
>         gram_schmidt(Q, range_product, atol=0, rtol=0, copy=False)
> ```
> 
> in `rand_la.rrf` correspond to Algorithm 2 of the cited paper.

If I remember correctly, this translates to different formulas, which, however, are mathematically equivalent. If you can't convince yourself that this is the case, I can take another look ...

**Reply by artpelling** [reply]
> There are several challenges: For instance, what do you do when you have multiple inner products around (say H1 and L2). If only one can be attached then the array would have to change space when you want to, e.g., orthogonalize w.r.t. the other product using `gram_schmidt`. (Or keep the cumbersome `product` argument.)

I'm not familiar with the concept of multiple inner products in one space. I guess we could stick with only allowing one `product` per space and in the mentioned case one could somehow create a new instance of the `VectorArray` such that it belongs to a space with the other product. Maybe using `with_`?

> If you have `VectorArrays` from the same (algebraic) space but with different inner products: what would their sum be?

Simply the sum. The `space` of the result could be chosen implicitly as the `space` of the left summand maybe.

> Mathematically most severe: if I write `project(A, W, V)` the meaning depends on whether `A` is actually a linear operator from a Hilbert space to itself, in which case we would need the take the inner products with the vectors in `W`, or a bilinear form, in which case `A` would map to the dual space and `project` would have to take the dual pairing. Currently, it is implicitly assumed in this case, that the result of `A.apply` is given w.r.t. to the dual basis s.t. the dual pairing would just be the Euclidean inner product of the DOF vectors. One motivation for attaching inner products to `VectorSpaces` would be that for most of pyMOR this would allow us to work completely 'coordinate-free', not making any assumptions about the basis representation of these vectors. (@lbalicki has a use case for this, for instance.) But then we would have to introduce the concept of duality. I actually started working on this, but ran out of time. Since then the question is in limbo.

As a system theory guy, these examples scenarios seem very rare to me. If this is not a thing regularly encountered in RB-world, maybe we could live with that. 

> I don't think this is inconsistent: You can only define the adjoint of an `Operator` when you know the inner products w.r.t. which you want to take the adjoint.
> I don't know about this. How would `source_product` and `range_product` affect the behavior of the operator's interface (except taking the adjoint)?

Forget about the adjoint, I was on the wrong path there.

> I'm not sure if that makes sense. What properties would you like the result to have? I think what you would want is a standard normal distribution of the coefficients w.r.t. an orthonormal basis of the Hilbert space. For that you would need to compute the inverse square root of `source_product` if I'm not mistaken. Since this is not so easy, `adaptive_rrf` pays `1/sqrt(lambda_min)` as a penalty for using the wrong basis. As far as I remember, when @andreasbuhr worked on this, he could not find a way to resolve this issue.

Yes, I came to a similar conclusion.

> If I remember correctly, this translates to different formulas, which, however, are mathematically equivalent. If you can't convince yourself that this is the case, I can take another look ...

I could convince myself (only experimentally though :wink:), thanks!
